analysis/src/python/utils/df_utils.py

```diff
@@ -0,0 +1,101 @@
+import os
+from pathlib import Path
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+import pandas as pd
+from pandarallel import pandarallel
+
+from analysis.src.python.utils.extension_utils import AnalysisExtension, get_restricted_extension
+from analysis.src.python.utils.xlsx_utils import read_df_from_xlsx, write_df_to_xlsx
+
+
+def _apply_to_row(row: pd.Series, column: str, func: Callable) -> pd.Series:
+    """ Apply `func` to data in `column` of dataframe's `raw`. """
+
+    copy_row = row.copy()
+    copy_row[column] = func(copy_row[column])
+    return copy_row
+
+
+def apply(df: pd.DataFrame, column: str, func: Callable) -> pd.DataFrame:
+    """ Apply `func` to  data in `column` of dataframe `df`. """
+
+    return df.apply(lambda row: _apply_to_row(row, column, func), axis=1)
+
+
+def parallel_apply(df: pd.DataFrame, column: str, func: Callable) -> pd.DataFrame:
+    """ Parallel apply `func` to  data in `column` of dataframe `df`. """
+
+    pandarallel.initialize(nb_workers=4)
+    return df.parallel_apply(lambda raw: _apply_to_row(raw, column, func), axis=1)
+
+
+def filter_df_by_iterable_value(df: pd.DataFrame, column: str, value: Iterable) -> pd.DataFrame:
+    return df.loc[df[column].isin(value)]
+
+
+def filter_df_by_single_value(df: pd.DataFrame, column: str, value: Any) -> pd.DataFrame:
+    return df.loc[df[column] == value]
+
+
+def drop_duplicates(df: pd.DataFrame, column: str) -> pd.DataFrame:
+    return df.drop_duplicates(column, keep='last').reset_index(drop=True)
+
+
+def rename_columns(df: pd.DataFrame, columns: Dict[str, str]) -> pd.DataFrame:
+    """ Rename columns of given dataframe `df`. """
+
+    return df.rename(columns=columns)
+
+
+def drop_columns(df: pd.DataFrame, columns: List[str]) -> pd.DataFrame:
+    """ Drop columns from given dataframe `df`. """
+
+    return df.drop(labels=columns, axis=1)
+
+
+def equal_df(expected_df: pd.DataFrame, actual_df: pd.DataFrame) -> bool:
+    return expected_df.reset_index(drop=True).equals(
```

Comment:
please, change the order - firstly check the empty condition and then next the dfs, because in Python all boolean check are lazy and check if the df is empty is faster